tests/test_h3.py

```diff
@@ -0,0 +1,1072 @@
+'''
+ Copyright (c) 2018 Uber Technologies, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+'''
+
+import pytest
+import unittest
+
+from h3 import h3
+from numbers import Number
+
+
+class TestH3Core(unittest.TestCase):
+    def shift_circular_list(self, start_element, elements_list):
+        # We shift the circular list so that it starts from start_element,
+        start_index = elements_list.index(start_element)
+        return elements_list[start_index:] + elements_list[:start_index]
+
+    def test_h3_is_valid(self):
+        self.assertTrue(
+            h3.h3_is_valid('85283473fffffff'),
+            'H3 Address is considered an address'
+        )
+        self.assertTrue(
+            h3.h3_is_valid('850dab63fffffff'),
+            'H3 Address from Java test also valid'
+        )
+        self.assertFalse(
+            h3.h3_is_valid('lolwut'),
+            'Random string is not considered an address'
+        )
+        self.assertFalse(
+            h3.h3_is_valid('5004295803a88'),
+            'H3 0.x Addresses are not considered valid'
+        )
+        for res in range(16):
+            self.assertTrue(
+                h3.h3_is_valid(h3.geo_to_h3(37, -122, res)),
+                'H3 Address is considered an address'
+            )
+
+    def test_geo_to_h3(self):
+        self.assertEqual(
+            h3.geo_to_h3(37.3615593, -122.0553238, 5), '85283473fffffff',
+            'Got the expected H3 address back'
+        )
+
+    def test_h3_get_resolution(self):
+        for res in range(16):
+            h3_address = h3.geo_to_h3(37.3615593, -122.0553238, res)
+            self.assertEqual(
+                h3.h3_get_resolution(h3_address), res,
+                'Got the expected H3 resolution back'
+            )
+
+    # note: changed from master
+    # note: not sure what the correct return value is here...
+    def test_silly_geo_to_h3(self):
+        out = h3.geo_to_h3(37.3615593 + 180.0, -122.0553238 + 360.0, 5)
+
+        expected_master_branch = '85283473fffffff'
+        expected_cython_branch = '85ca2d53fffffff'
+
+        assert out == expected_master_branch
```

Comment:
This is the last test that's failing for me.

The current `master` branch expects `'85283473fffffff'`.
The new `cython` branch returns `'85ca2d53fffffff'`.

Any idea which one is the correct result? @isaacbrodsky @nrabinowitz @dfellis?

Reply:
Could this have anything to do with me dropping the ["mercator" transform in the last commit](https://github.com/uber/h3-py/commit/963241447bba1151958199c6ac30d726c34a2345#diff-1df158fef384a0b5e893d1ef4a14c773L13-L20)?

We did that because we decided that should have been handled by the C library. We sure that's the case?

Reply:
The first index (master) is what the C library gives, too. I also tested an older version of the core library and got the same result. However, the "mercator" function seems to be wrapping the north pole (90 degrees) around to the south pole (-90 degrees), which I'm not sure makes sense. I think this behavior may be correct.

Reply:
The JS lib gives `85ca2d53fffffff` (Cython branch). Not sure why this would be different from the C library :/.

Reply:
It's a lat/lng normalization issue. I'm pretty sure `85283473fffffff` is the correct cell here.

I don't know why it was called that, but the `mercator()` function that we previously removed just tried to ensure that `lat` is in the interval `[-90, 90]` (should it be half-open?) and `lng` is in `[-180, 180]`.

The function would thus transform:

```
(37.3615593 + 180.0, -122.0553238 + 360.0) -> (37.3615593, -122.0553238)
```

Weird that the C library seems to do this normalization correctly, but the Cython and JS libs do not. Especially since they're just calling the C lib.

@isaacbrodsky, how did you test this with the C lib? Any chance there's some pre-processing code that is normalizing the coordinates?

Reply:
Maybe we want a function like this?

```python
def normalize(lat, lng):
    """ Helper to coerce lat/lng range
    
    lat in [-90, +90]
    lng in [-180, +180]
    
    """
    while lat > +90:
        lat -= 180
    
    while lat < -90:
        lat += 180

    while lng > +180:
        lng -= 360
        
    while lng < -180:
        lng += 360

    return lat, lng
```

We could also just do some modulus operation, but I think this makes the logic clear.

Reply:
Alternatively, I guess we could also just throw an error if the inputs are outside of expected bounds?

Reply:
I tested using the command line `geoToH3` program. It just converts the coordinates to radians and feeds them into heoToH3. I don't think that wrapping to the south pole makes sense.

Reply:
Right, but wrapping back the other direction also doesn't make sense without changing the longitude at the same time. If we treat the coordinates as a vector from `(0, 0)` then getting to a latitude of 91 degrees really implies a latitude of 89 degrees while also adding 180 degrees to the longitude and clamping again, right?

Reply:
I'll resolve this for now by adding back in the `mercator()` function, so that the Cython branch agrees with `master` and the C library.

Kicking the discussion of what we should **really** do to https://github.com/uber/h3/issues/287.

Reply:
Actually, I've switched my opinion on this. Notes here: https://github.com/uber/h3/issues/287